File: word_chain.pl
use strict;
use warnings;

package WordChain;

sub new {
    my ($class, $n, $d_ref) = @_;
    my $self = {
        n      => $n,
        d      => $d_ref,
        alive  => [(1) x $n],
        bef    => "",
        p      => 0
    };
    return bless $self, $class;
}

sub wordChain {
    my ($self, $s) = @_;
    my $n = $self->{n};
    if (exists $self->{d}->{$s} && ($self->{bef} eq "" || substr($s, 0, 1) eq substr($self->{bef}, -1)) && substr($s, -1) ne 'z') {
        delete $self->{d}->{$s};
        $self->{bef} = $s;
    } else {
        $self->{alive}->[$self->{p}] = 0;
        $self->{bef} = "";
    }
    do {
        $self->{p} = ($self->{p} + 1) % $n;
    } while (!$self->{alive}->[$self->{p}]);
}

sub aliveCount {
    my ($self) = @_;
    return scalar grep { $_ } @{$self->{alive}};
}

sub isAlive {
    my ($self, $i) = @_;
    return $self->{alive}->[$i];
}

package main;

my ($n, $k, $m);
chomp(my $input = <STDIN>);
($n, $k, $m) = split ' ', $input;

my %d;
while ($k--) {
    chomp(my $word = <STDIN>);
    $d{$word} = 1;
}

my $wordChain = WordChain->new($n, \%d);
while ($m--) {
    chomp(my $word = <STDIN>);
    $wordChain->wordChain($word);
}

print $wordChain->aliveCount(), "\n";
for my $i (0 .. $n - 1) {
    if ($wordChain->isAlive($i)) {
        print $i + 1, "\n";
    }
}

File: word_chain.cs
using System;
using System.Linq;
using System.Collections.Generic;

class WordChain
{
    int n;
    HashSet<string> d;
    bool[] alive;
    string bef;
    int p;

    public WordChain(int n, HashSet<string> d)
    {
        this.n = n;
        this.d = d;
        alive = new bool[n];
        for (int i = 0; i < n; i++) alive[i] = true;
        bef = string.Empty;
        p = 0;
    }

    public void Chain(string s)
    {
        if (d.Contains(s) && ("".Equals(bef) || s.First() == bef.Last()) && s.Last() != 'z')
        {
            d.Remove(s);
            bef = s;
        }
        else
        {
            alive[p] = false;
            bef = string.Empty;
        }
        do
        {
            p = -~p % n;
        } while (!alive[p]);
    }

    public int AliveCount()
    {
        return alive.Sum(b => b ? 1 : 0);
    }

    public bool IsAlive(int i)
    {
        return alive[i];
    }
}

public class Program
{
    public static void Main()
    {
        int[] nkm = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
        int n = nkm[0];
        int k = nkm[1];
        int m = nkm[2];
        HashSet<string> d = new HashSet<string>();
        while (k-- > 0) d.Add(Console.ReadLine());
        WordChain wordChain = new WordChain(n, d);
        while (m-- > 0) wordChain.Chain(Console.ReadLine());
        Console.WriteLine(wordChain.AliveCount());
        for (int i = 0; i < n; i++) if (wordChain.IsAlive(i)) Console.WriteLine(-~i);
    }
}
